eras/babbage/test-suite/cddl-files/babbage.cddl

```diff
@@ -0,0 +1,404 @@
+block =
+  [ header
+  , transaction_bodies         : [* transaction_body]
+  , transaction_witness_sets   : [* transaction_witness_set]
+  , auxiliary_data_set         : {* transaction_index => auxiliary_data }
+  , invalid_transactions       : [* transaction_index ]
+  ]; Valid blocks must also satisfy the following two constraints:
+   ; 1) the length of transaction_bodies and transaction_witness_sets
+   ;    must be the same
+   ; 2) every transaction_index must be strictly smaller than the
+   ;    length of transaction_bodies
+
+transaction =
+  [ transaction_body
+  , transaction_witness_set
+  , bool
+  , auxiliary_data / null
+  ]
+
+transaction_index = uint .size 2
+
+header =
+  [ header_body
+  , body_signature : $kes_signature
+  ]
+
+header_body =
+  [ block_number     : uint
+  , slot             : uint
+  , prev_hash        : $hash32 / null
+  , issuer_vkey      : $vkey
+  , vrf_vkey         : $vrf_vkey
+  , vrf_result       : $vrf_cert ; New, replaces nonce_vrf and leader_vrf
+  , block_body_size  : uint
+  , block_body_hash  : $hash32 ; merkle triple root
+  , operational_cert
+  , protocol_version
+  ]
+
+operational_cert =
+  ( hot_vkey        : $kes_vkey
+  , sequence_number : uint
+  , kes_period      : uint
+  , sigma           : $signature
+  )
+
+protocol_version = (uint, uint)
+
+transaction_body =
+ { 0 : set<transaction_input>    ; inputs
+ , 1 : [* transaction_output]
+ , 2 : coin                      ; fee
+ , ? 3 : uint                    ; time to live
+ , ? 4 : [* certificate]
+ , ? 5 : withdrawals
+ , ? 6 : update
+ , ? 7 : auxiliary_data_hash
+ , ? 8 : uint                    ; validity interval start
+ , ? 9 : mint
+ , ? 11 : script_data_hash
+ , ? 13 : set<transaction_input> ; collateral inputs
+ , ? 14 : required_signers
+ , ? 15 : network_id
+ , ? 16 : transaction_output     ; collateral return; New
+ , ? 17 : coin                   ; total collateral; New
+ , ? 18 : set<transaction_input> ; reference inputs; New
+ }
+
+required_signers = set<$addr_keyhash>
+
+transaction_input = [ transaction_id : $hash32
+                    , index : uint
+                    ]
+
+transaction_output = legacy_Transaction_output / post_alonzo_transaction_output ; New
+
+legacy_transaction_output =
+  [ address
+  , amount : value
+  , ? datum_hash : $hash32
+  ]
+
+; Note that in the post_alonzo_transaction_output, we do not allow
+; a transaction output to include both a datum hash and a datum.
+; In other words, keys 2 and 3 are mutually exclusive.
+post_alonzo_transaction_output =
+     { 0 => address
+     , 1 => value
+     , ? 2 => datum_hash
+     , ? 3 => datum      ; New; inline datum
+     , ? 4 => script     ; New; script reference
+     }
```

Comment:
i like this 👍, how does deserializing the TxOut with cborg work though? do we decide how to branch by `peekTokenType`? cborg doesn't seem to have great machinery for deserializing choices

Reply:
Unfortunately I don't think that there is a way to handle it besides branching on `peekTokenType`.